ls1intum/Artemis
# `Plagiarism checks`: Display text submissions even though similarity highlighting does not work

### Checklist
#### General



#### Client

### Motivation and Context


We had a case when server returned PlagiarismComparison matches which has `startA`/`startB` indices greater than the length of the length of `submissionA.elements`/`submissionB.elements`. In this and sileven though similarity highlighting does not work, the users should be able to see both text submissions. This allows them to at least compare the submissions manually.

### Description

This PR adds a fail-soft fix that will display text submissions even though matches cannot be mapped to elements correctly due to out-of-bounds indices.

Also, this PR adds "overall similarity" percentage above the split view for both instructors and students.

Additionally, resolves https://github.com/ls1intum/Artemis/issues/5403 with a new button for navigating to Exercise Plagiarism Detection page from the Plagiarism Case Instructor Detail view.

Finally, this PR addresses the root cause of out-of-bounds indices, which was caused by the swap operation between `submissionA` and `submissionB` of `PlagiarismComparison`. It did not swap `PlagiarismComparison.matches[].startA` and `startB` fields so the indices did not match.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Text Exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. Make two similar text submissions with both students
  To reproduce the problematic "Before" case in the screenshot:
    - One submission must be longer than the other.
    - There should be similarities starting near the end of the submissions (You can just copy/paste a short paragraph at the end of both submissions).
5. As the instructor, set the due date of the exercise to now and run plagiarism check from Exercise Detail page > Check Plagiarism page.
6. Notify a student about the plagiarism.
7. With both the instructor and the notified student, check if the exercise detail page shows both submission in the split view and highlighted similarities are correct.
8. Repeat the same steps for modeling and programming exercises.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| plagiarism-case-instructor-detail-view.component.ts | 72.91% | 88.23% |
| plagiarism-split-view.component.ts | 90.32% | 94.91% |
| text-submission-viewer.component.ts | 79.31% | 94.59% |

### Screenshots



Before:
![image](https://user-images.githubusercontent.com/18427209/182453901-c29dfa41-9946-4d27-ba74-53724e778040.png)

After handling out-of-bounds indices:
![image](https://user-images.githubusercontent.com/18427209/182495954-b7647623-1d4d-446d-9580-d9fa74fa1ae1.png)

After fixing submission swapping:
![image](https://user-images.githubusercontent.com/18427209/182600983-acc279ef-1964-4e11-8d70-45ceec4456ed.png)

New button for navigating to Exercise Plagiarism Detection page:
![image](https://user-images.githubusercontent.com/18427209/182658211-a6b13890-55a1-4ed3-83d5-c4e9c3ad9feb.png)
